### PR TITLE
executor: checking chunk is full precedes filtering

### PR DESCRIPTION
**Pull request by ikenchina**
### What problem does this PR solve?

Consider a corner case : If chunk is full(limit), continuously check filter conditions and there is not matched row is meaningless.

### What is changed and how it works?

What's Changed: checking chunk is full precedes checking filter conditions.
Penalty is that need to check chunk every time, it cost more CPU but it is negligible.

How it Works:

```
CREATE TABLE test_sel_exec(
id INT AUTO_INCREMENT PRIMARY KEY,
status TINYINT NOT NULL
);

INSERT INTO test_sel_exec(status) values(0);
INSERT INTO test_sel_exec(status) values(0);
INSERT INTO test_sel_exec(status) values(0);

INSERT INTO test_sel_exec(status) values(1);
INSERT INTO test_sel_exec(status) values(1);
INSERT INTO test_sel_exec(status) values(1);
INSERT INTO test_sel_exec(status) values(1);

SELECT * FROM test_sel_exec where status=0 order by id limit 3;
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Fix : avoid continuously scan mismatched rows if chunk is full

### CODE REVIEW

**Comment by ikenchina** [comment]
@qw4990 can you help to review?

**Comment by guo-shaoge** [comment]
Normally we push `Selection` to TiKV, so I am not sure in which scenarios this fix will improve performance and how much it will improve. (You can do more investigations if you like.)

But it still looks good to me.

File: executor/executor.go [hunk 1]
```diff
@@ -1251,12 +1251,14 @@ func (e *SelectionExec) Next(ctx context.Context, req *chunk.Chunk) error {
 
 	for {
 		for ; e.inputRow != e.inputIter.End(); e.inputRow = e.inputIter.Next() {
-			if !e.selected[e.inputRow.Idx()] {
-				continue
-			}
 			if req.IsFull() {
```
Comment (lzmhhh123) [inline executor/executor.go:1254]: Maybe a more reasonable way is to move the check below `req.AppendRow`?

Comment (ikenchina) [reply executor/executor.go:1254]: No, logical will be wrong.

**Review by lzmhhh123 — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**